### PR TITLE
test ALL parsed inventory sources result instead of relying on ONE su…

### LINKED ISSUE #45382: possible unreliable aggregated inventory when parsing multiple inventories using inventory plugins and when any of them raise an error
##### SUMMARY
assuming the following:
- using ansible custom inventory plugins
- one inventory plugin (good.py) returning results correctly
- the other one (bad.py) raising an error during parsing
- enabling unparsed_is_failed

considering the following scenarios:
- running ansible-inventory against good.py returns no error (rc==0)
- running ansible-inventory against bad.py returns an error (rc==1)
- running ansible-inventory against bad.py AND good.py (multiple inventories support) returns NO error (rc==0)

I'm assuming the inventory step as atomic. At least we should have the option to request it (IMHO, unparsed_is_failed should be made for). Otherwise, you never know what incomplete inventory can have as side effect against your targets.

##### COMMENTS
using 2 inventory plugins is just for test POC convenience. We could only use one, parsing different source of truth, one providing reliable content but not the other one and get the same behavior

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible inventory plugin

##### ANSIBLE VERSION
```
ansible-inventory 2.6.4
  config file = /home/flabatut/test/ansible.cfg
  configured module search path = [u'/home/flabatut/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/flabatut/test/venv/lib/python2.7/site-packages/ansible
  executable location = /home/flabatut/test/venv/bin/ansible-inventory
  python version = 2.7.14 (default, Nov  3 2017, 10:55:25) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

##### CONFIGURATION

```
DEFAULT_INVENTORY_PLUGIN_PATH(/home/flabatut/test/ansible.cfg) = [u'/home/flabatut/test/inventory_plugins']
INVENTORY_UNPARSED_IS_FAILED(/home/flabatut/test/ansible.cfg) = True
```

##### OS / ENVIRONMENT
Linux test 4.13.16-302.fc27.x86_64 #1 SMP Thu Nov 30 15:33:36 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
Fedora release 27 (Twenty Seven)

## assuming the following file structure:
```
.
├── venv
|    └── ...
├── ansible.cfg
├── bad.yml
├── good.yml
└── inventory_plugins
    ├── bad.py
    ├── bad.pyc
    ├── good.py
    └── good.pyc
```

## creating root folder
```
mkdir -p test/inventory_plugins
cd test
```

## installing virtualenv/ansible
```
virtualenv venv
pip install --upgrade pip
pip install ansible
. venv/bin/activate
```

## creating ansible.cfg
```
cat <<EOF> ansible.cfg
[defaults]
inventory_plugins = ./inventory_plugins

[inventory]
unparsed_is_failed = true
EOF
```

## creating inventory plugins
```
cat <<EOF> ./inventory_plugins/good.py
from __future__ import (absolute_import, division, print_function)
from ansible.plugins.inventory import BaseInventoryPlugin
__metaclass__ = type

class InventoryModule(BaseInventoryPlugin):
    NAME = 'good'
    def __init__(self):
        super(InventoryModule, self).__init__()

    def parse(self, inventory, loader, path, cache=False):
        """
        main function
        """
        super(InventoryModule, self).parse(inventory, loader, path)
EOF
```

```
cat <<EOF> ./inventory_plugins/bad.py
from __future__ import (absolute_import, division, print_function)
from ansible.plugins.inventory import BaseInventoryPlugin
from ansible.errors import AnsibleParserError
__metaclass__ = type

class InventoryModule(BaseInventoryPlugin):
    NAME = 'bad'
    def __init__(self):
        super(InventoryModule, self).__init__()

    def parse(self, inventory, loader, path, cache=False):
        """
        main function
        """
        super(InventoryModule, self).parse(inventory, loader, path)
        raise AnsibleParserError('error detected')
EOF
```

## creating  inventory plugin config files
```
cat <<EOF> good.yml
plugin: good
EOF
```

```
cat <<EOF> bad.yml
plugin: bad
EOF
```

##### EXPECTED RESULTS
when providing multiple inventories to ansible, if any of those inventories returns any error, we should prevent ansible to perform any modification using an unreliable and incomplete aggregated inventory

## testing the good inventory plugin, no error expected
```
$ ANSIBLE_DEBUG=1 ansible-inventory --list -i good.yml >/dev/null ; echo $?
0
```

## testing the bad inventory plugin returns an error
```
$ ANSIBLE_DEBUG=1 ansible-inventory --list -i bad.yml >/dev/null ; echo $?
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with yaml plugin: Plugin configuration YAML file, not YAML inventory
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with ini plugin: /home/flabatut/test/bad.yml:1: Expected key=value host variable assignment, got: bad
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with auto plugin: error detected
 [WARNING]: Unable to parse /home/flabatut/test/bad.yml as an inventory source
ERROR! No inventory was parsed, please check your configuration and options.
1
```
what's interesting here is:
```
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with auto plugin: error detected
```
caused by the raise instruction in the inventory plugin, which is expected

## testing both inventories aggregated by ansible returns NO errors (just warnings) AND an incomplete inventory
```
$ ANSIBLE_DEBUG=1 ansible-inventory --list -i bad.yml -i good.yml >/dev/null ; echo $?
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with yaml plugin: Plugin configuration YAML file, not YAML inventory
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with ini plugin: /home/flabatut/test/bad.yml:1: Expected key=value host variable assignment, got: bad
 [WARNING]:  * Failed to parse /home/flabatut/test/bad.yml with auto plugin: error detected
 [WARNING]: Unable to parse /home/flabatut/test/bad.yml as an inventory source
0
```

Considering an inventory returning a few thousand lines in JSON format, warnings can't be catched visually. Apart from that, by experience, when such issue occurs, the aggregated and returned result becomes incomplete and unreliable

### PR DESCRIPTION
**Pull request by ghost**
…ccessfull only

##### SUMMARY
Fixes #45382

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ansible inventory plugin

##### ANSIBLE VERSION
```
ansible-inventory 2.6.4
  config file = /home/flabatut/test/ansible.cfg
  configured module search path = [u'/home/flabatut/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/flabatut/test/venv/lib/python2.7/site-packages/ansible
  executable location = /home/flabatut/test/venv/bin/ansible-inventory
  python version = 2.7.14 (default, Nov  3 2017, 10:55:25) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

affecting also 2.8.0.dev0

##### ADDITIONAL INFORMATION
See https://github.com/ansible/ansible/issues/45382 for detailed steps used to reproduce the issue

### CODE REVIEW

File: lib/ansible/inventory/manager.py [hunk 1]
```diff
@@ -195,23 +195,25 @@ def parse_sources(self, cache=False):
...
                     source = unfrackpath(source, follow=False)
                 parse = self.parse_source(source, cache=cache)
-                if parse and not parsed:
-                    parsed = True
+                all_parsed_results.append(parse)
 
-        if parsed:
+        # ensure all inventories parsed returned true
+        print(all_parsed_results)
+        if all(all_parsed_results):
```
Comment (s-hertel) [inline lib/ansible/inventory/manager.py]: If there have been parsed results and C.INVENTORY_UNPARSED_IS_FAILED is not True I'd expect post-processing to continue.

Comment (ghost) [reply lib/ansible/inventory/manager.py]: indeed @s-hertel , perhaps append after line 216, the same the post-processing instruction declared above ? 
self._inventory.reconcile_inventory()

Comment (s-hertel) [reply lib/ansible/inventory/manager.py]: @flabatut Sure. You could also do something like I did in #45387. Lots of ways.

File: lib/ansible/inventory/manager.py [hunk 2]
```diff
@@ -206,15 +206,22 @@ def parse_sources(self, cache=False):
...
 
-        # ensure all inventories parsed returned true
+        # ensure all parsed inventories returned true
         if all(all_parsed_results):
             # do post processing
             self._inventory.reconcile_inventory()
         else:
             if C.INVENTORY_UNPARSED_IS_FAILED:
-                raise AnsibleError("No inventory was parsed or errors detected, please check your configuration and options.")
+                raise AnsibleError("Errors detected during parsing, please check your configuration and options.")
```
Comment (s-hertel) [inline lib/ansible/inventory/manager.py]: The word "inventory" should probably still be part of the error message just to make it more specific what sort of configuration/options to check.

**Comment by s-hertel** [comment]
@flabatut Oh, it looks like there are two separate options. See if `any_unparsed_is_failed` does what you need. It looks like `unparsed_is_failed` is for every single potential inventory source failing to be parsed, rather than only a subset.

**Pull request closed without merging**